### PR TITLE
Speed git downloads by using cache dir as reference

### PR DESCRIPTION
**Pull request by luisfaceira**
Proposed solution for #1323

When downloading a git repository, it checks if that repository was already cached under cache-vcs-dir and if it was, use it as a "--reference".

By using this option, when downloading objects, git checks if those objects exist in that local mirror first, greatly reducing the network bandwidth usage, speeding up composer when reusing a package across multiple projects. See https://git-scm.com/docs/git-clone

The cache was already being created, when composer analyzes the dependencies of the source repo, so I only needed to check if the cache dir already existed.

Does anyone see any drawbacks of using this approach?

I tested manually on multiple situations (new repo, update, install, repo with new commits, etc.).

BTW, the original motivation to this was that trying to find a work-around to #3542 different than setting oauth tokens (which bring shared-security issues), I initially htought of using --prefer-source and was surprised that no cache was being used. Maybe with this, the performance penalty of using --prefer-source reduces enough for it to be an alternative to downloads from github api?

------------
> This pull request was made during the Open Source Day @ [beubi](http://beubi.com) :green_heart:

### CODE REVIEW

File: src/Composer/Downloader/GitDownloader.php [hunk 1]
```diff
@@ -41,10 +41,12 @@ public function doDownload(PackageInterface $package, $path, $url)
     {
         GitUtil::cleanEnv();
         $path = $this->normalizePath($path);
+        $cachePath = $this->config->get('cache-vcs-dir').'/'.preg_replace('{[^a-z0-9.]}i', '-', $url).'/';
+        $cacheOptions = file_exists($cachePath) ? $cacheOptions = '--reference '.$cachePath.' ' : '';
```
Comment (staabm) [inline src/Composer/Downloader/GitDownloader.php]: Does it miss a `ProcessExecutor::escape()` to not error on some special chars?

**Comment by staabm** [comment]
Do you have some number before/after the patch?

**Comment by luisfaceira** [comment]
@staabm I do. I was developing with a single git repo included (~90M download) and got the following average times:
* first composer install: 14 to 18s
* following composer install (after removing lock and vendor): 1.5 to 3s
* update with a simple additional commit to be downloaded: 9 to 10s

**Comment by staabm** [comment]
How long took it without your patch?

**Comment by luisfaceira** [comment]
Without the patch it takes about the same time that the first composer install, but in every execution, not just the first one, because it was always downloading the 90M repo.

File: src/Composer/Downloader/GitDownloader.php (hunk 1 above)
Comment (luisfaceira) [inline src/Composer/Downloader/GitDownloader.php]: Hmmm maybe it does, I'm not sure.

Special chars where, in the setting of "cache-vcs-dir" ?
The preg_replace applied was extracted from similar usages in the VcsDrivers, thought that would be enough.

Do you think I need to add that ProccessExecuter::escape() ?

Comment (staabm) [inline src/Composer/Downloader/GitDownloader.php]: If it works in other drivers I guess it is ok

File: src/Composer/Config.php [hunk 2]
```diff
@@ -49,6 +49,7 @@ class Config
         'platform' => array(),
         'archive-format' => 'tar',
         'archive-dir' => '.',
+        'home' => '$HOME'
```
Comment (alcohol) [inline src/Composer/Config.php]: Not sure how this change relates to your PR?

File: src/Composer/Downloader/GitDownloader.php (hunk 1 above)
Comment (alcohol) [inline src/Composer/Downloader/GitDownloader.php]: If you check line 51 (now 53) you will note that all paths are escaped using `ProcessExecutor::escape()`. The same should be done for the `--reference` argument value (`$cachePath`).

File: src/Composer/Config.php (hunk 2 above)
Comment (luisfaceira) [inline src/Composer/Config.php]: When I added the usage of $this->config->get('cache-vcs-dir') , the GitDownloaderTest started failing with notices, because in that context, the config wasn't fully initialized, and the 'home' config was missing. Setting it in the default configs solved it.

Setting a default 'home' config that is automatically

File: src/Composer/Downloader/GitDownloader.php (hunk 1 above)
Comment (luisfaceira) [inline src/Composer/Downloader/GitDownloader.php]: Ok. I haven't yet understood completely in what situation this might be necessary in this specific case where I'm getting a dir from conf, but better be safe than sorry :) will do the change right away!

File: src/Composer/Config.php (hunk 2 above)
Comment (alcohol) [inline src/Composer/Config.php]: @seldaek any idea if this might have adverse effects?

**Comment by cs278** [comment]
I think you should use `--dissociate --reference /path/to/cache` otherwise the cloned repository is dependant on the cache directory existing, if you remove the cache directory the repo in vendor will break. The downside is the network I/O saving will only apply to initial clones.

**Comment by luisfaceira** [comment]
Hmmm I was convinced that --reference would set an "alternate" source, and therefore, if it was not present, it would proceed normally. But after digging after your question, @cs278, it seems you're right, its safer to use --dissociate (also see #3722 ).
I will add it and try to see if I can reference it again temporarily on updates as well.

File: src/Composer/Config.php (hunk 2 above)
Comment (stof) [inline src/Composer/Config.php]: Well, this would not work. You would get ``$HOME/cache`` as cache dir, instead of resolving the env variable. IMO, the right fix is making sure the config you pass to the test is initialized properly

Comment (luisfaceira) [inline src/Composer/Config.php]: Can you please examplify in what situations won't it work, @stof, I'm not understanding.
I haven't debugged it in such detail, but from looking at [Composer/Config.php#L181](https://github.com/luisfaceira/composer/blob/faster-git-clones-using-cache/src/Composer/Config.php#L181) I would expect you to get the full cache dir expanded from your environment such as `/home/myuser/.composer/cache` and not simply `$HOME/cache` as you said.

And if it does not work, what would you suggest to be a proper initialization in the test - explicitly setting in runtime the home config from the home environment, when initializing the Config class?

Comment (luisfaceira) [inline src/Composer/Config.php]: Here's the issue that I think that made it possible to use `$HOME`: #3060

Comment (stof) [inline src/Composer/Config.php]: but ``$config->get('home')`` would be broken in your case though

**Comment by stof** [comment]
Note that this will only affect repositories added as git repositories in your composer.json. Any git repository for a package loaded from Satis or Packagist won't be in your vcs dir cache

File: src/Composer/Config.php (hunk 2 above)
Comment (luisfaceira) [inline src/Composer/Config.php]: Yes, I think you're right about the `$config->get('home')`, it won't expand properly the usage of $HOME inside, in the current implementation of that feature.

I'm not sure if I should improve (possibly in another PR?) what has been implemented in 449f8165ef64829dc34be836235ceb74d9227dd7 by @seldaek and make $HOME be interpreted also when used in the 'home' config (although it's not called "-dir" it is a dir config, after all, is it not?), or if I should instead force setting the home config value when creating the object for the GitDownloaderTest.

What do you think?

**Comment by luisfaceira** [comment]
@stof I was aiming for that case, of git repositories added in composer.json. I'm not sure if I can squeeze the time in the weekend, but I'll try to expand this to cover more cases, and I'll also try to look into the option of using shallow cloning to speed things up even more (in separate branches).

File: src/Composer/Config.php (hunk 2 above)
Comment (luisfaceira) [inline src/Composer/Config.php]: After thinking this through and all the debate that followed this change, I decided that it would be a lot safer to revert it, and instead change the test to set a temp dir as home when creating the Donwloader Mock.

File: tests/Composer/Test/Downloader/GitDownloaderTest.php [hunk 3]
```diff
@@ -26,6 +26,10 @@ protected function getDownloaderMock($io = null, $config = null, $executor = nul
         if (!$config) {
             $config = new Config();
         }
+        if (!$config->has('home')) {
```
Comment (staabm) [inline tests/Composer/Test/Downloader/GitDownloaderTest.php]: Should we do a cleanup of this directory in the test teardown?

Comment (luisfaceira) [inline tests/Composer/Test/Downloader/GitDownloaderTest.php]: My change only ensures that an home config is set, because there was some pre-existing code in [Config](https://github.com/composer/composer/blob/master/src/Composer/Config.php#L222) that was expecting it to exist.

I've only set it up as a dir inside /tmp/ to make sure that it would not break if something in the future actually needs a valid dir, and used the exact same technique to generate a random name that was being used by others in this test, for consistency. But, at least with the current tests, the home dir isn't actually created, so no teardown related to my change is necessary, at least for now. 

However, there are a bunch of dir and files that tests create (both in the GitDownloaderTest and other tests), at least in my /tmp/ after running phpunit I find a lot of leftovers:
> cmptest-2003111fcf4331c420d8c61ef429410f
> cmptest-3c7103f6ac29a4865b351466a4a730da
> cmptest-a71ac06271f7e35caa3c05f03206e459
> composer-test
> composer_testdir
> composer_test_file
> composer-xz-test-vendor
> composer-zip-test-vendor`

IMHO, files created inside /tmp/ do not absolutely require teardown, that would be a nice-to-have. But if it's considered to be relevant, maybe we should open an issue for it?

Comment (alcohol) [inline tests/Composer/Test/Downloader/GitDownloaderTest.php]: Agreed, it's not relevant to this PR. Don't worry about it.

Comment (staabm) [inline tests/Composer/Test/Downloader/GitDownloaderTest.php]: > But if it's considered to be relevant, maybe we should open an issue for it?

Having this information in a separate Issue to be fixed in a separate PR would be great IMO

Comment (luisfaceira) [inline tests/Composer/Test/Downloader/GitDownloaderTest.php]: opened #4694 to do so in all tests that currently create dirs

**Comment by hkdobrev** [comment]
We are using `config.preferred-install` to install our own organisation packages from source and this would be a very welcome change. Are there any updates on this?

**Comment by Seldaek** [comment]
Nope, stlil have to give it a try

**Comment by Seldaek** [comment]
Closing in favor of #5384 which includes these changes
